scripts/finalizeCCTPMessages.ts

```diff
@@ -0,0 +1,309 @@
+ * --network linea-goerli \
+ * ```
+ * Environment variables:
+ * - `RECEIVE_MESSAGES_ON`: Which messages to claim. Either `l1` or `l2`.
+ * Flags:
+ * - `--network`: The L2 network, i.e. `linea-goerli` or `linea`.
+ *
+ */
+async function main() {
+  const receiveMessagesOn = process.env.RECEIVE_MESSAGES_ON || "l2";
+  const l1BlockLookback = parseInt(process.env.BLOCK_LOOKBACK || "8640");
+  const l2BlockLookback = parseInt(process.env.BLOCK_LOOKBACK || "17380");
+
+  const l1ChainId = parseInt(await hre.companionNetworks.l1.getChainId());
+  const l2ChainId = parseInt(await hre.getChainId());
+  const l1Provider = new ethers.providers.JsonRpcProvider(
+    getNodeUrl(l1ChainId === 1 ? "mainnet" : "goerli", true, l1ChainId)
+  );
+  const l2Provider = ethers.provider;
+  const l1Signer = ethers.Wallet.fromMnemonic((hre.network.config.accounts as any).mnemonic).connect(l1Provider);
+  const [l2Signer] = await ethers.getSigners();
+
+  console.log("\nL1 chain ID:", l1ChainId);
+  console.log("L2 chain ID:", l2ChainId);
+  console.log("Signer address:", l1Signer.address);
+
+  // Get relevant src events
+  const srcEvents =
+    receiveMessagesOn === "l1"
+      ? await getL2SrcEvents(l2ChainId, l2Provider, l2Signer, l2BlockLookback)
+      : await getL1SrcEvents(l1ChainId, l1Provider, l1Signer, l1BlockLookback);
+
+  // Get relevant message hashes and bytes
+  const uniqueTxHashes = new Set(srcEvents.map((event) => event.transactionHash));
+  const messageHashesAndBytes = await parseMessageHashesAndBytes(
+    Array.from(uniqueTxHashes),
+    receiveMessagesOn === "l1" ? l2Provider : l1Provider
+  );
+  const relevantMessageHashesAndBytes = messageHashesAndBytes.filter(
+    (messageHashAndBytes) =>
+      messageHashAndBytes.destinationDomain === CIRCLE_DOMAIN_IDs[receiveMessagesOn === "l1" ? l1ChainId : l2ChainId]
+  );
+  console.log(`\nParsed ${relevantMessageHashesAndBytes.length} relevant 'MessageSent' events`);
+
+  if (relevantMessageHashesAndBytes.length === 0) {
+    console.log("No relevant messages to receive");
+    return;
+  }
+
+  // Request attestations
+  console.log("\nRequesting attestations...");
+  const attestations = [];
+  for (const messageHashAndBytes of relevantMessageHashesAndBytes) {
+    const attestation = await requestAttestation(messageHashAndBytes.messageHash);
+    attestations.push(attestation);
+  }
+
+  // Receive messages
+  const messageTransmitter = new ethers.Contract(
+    receiveMessagesOn === "l1"
+      ? L1_ADDRESS_MAP[l1ChainId].cctpMessageTransmitter
+      : L2_ADDRESS_MAP[l2ChainId].cctpMessageTransmitter,
+    messageTransmitterAbi,
+    receiveMessagesOn === "l1" ? l1Signer : l2Signer
+  );
+
+  console.log(`\nReceiving messages on ${receiveMessagesOn.toUpperCase()}`, {
+    messageTransmitter: messageTransmitter.address,
+    chainId: receiveMessagesOn === "l1" ? l1ChainId : l2ChainId,
+  });
+  for (const [i, messageHashAndBytes] of relevantMessageHashesAndBytes.entries()) {
```

Comment:
Should we do this in parallel?

Reply:
Because I call a [write-method later in the loop](https://github.com/across-protocol/contracts-v2/pull/394/files#diff-104d868797f2bf90e2109e47b8cd64924323dbc433016f954ddbb94e09ba8465R158), parallelizing wouldn't work directly. I guess, I would have to use a multicall contract for that? But again the same reasoning as above: for debugging and the scope of this helper script, I think we don't need to parallelize.